### PR TITLE
[UX] Overlay improvements

### PR DESCRIPTION
**Pull request by BrettCleary**
Expecting some comments on this one :) 

See https://github.com/HyperPlay-Gaming/hyperplay-overlay/pull/3 for more discussion on the overlay parts

Also `getHandleFunction` could replace our `ipcRenderer.on` preload methods. It accomplishes the same goals with much less boilerplate

All submodule PR's:
- https://github.com/HyperPlay-Gaming/hyperplay-overlay/pull/3
- https://github.com/HyperPlay-Gaming/hyperplay-proxy-server/pull/25
- https://github.com/HyperPlay-Gaming/hyperplay-extension-helper/pull/31

---

Use the following Checklist if you have changed something on the Backend or Frontend:

- [ ] Tested the feature and it's working on a current and clean install.
- [ ] Tested the main App features and they are still working on a current and clean install. (Login, Install, Play, Uninstall, Move games, etc.)
- [ ] Created / Updated Tests (If necessary)
- [ ] Created / Updated documentation (If necessary)

### CODE REVIEW

**Comment by nyghtstalker** [comment]
- Clicking the wallet and MM buttons on top triggers many errors
- The wallet is connected yet still shows as not connected
- The MM modal stays open and can't be collapsed even after connecting
- Blank white screen when launching a game 

image

image

**Comment by BrettCleary** [comment]
> * Clicking the wallet and MM buttons on top triggers many errors
> * The wallet is connected yet still shows as not connected
> * The MM modal stays open and can't be collapsed even after connecting
> * Blank white screen when launching a game
> 
> image
> 
> image

It would appear you are not on the correct `hyperplay-overlay` submodule branch @nyghtstalker

File: src/backend/launcher.ts [hunk 1]
```diff
@@ -789,8 +789,12 @@ async function callRunner(
...
     })
 
-    if (gameInfo && gameInfo.runner === 'hyperplay')
-      openOverlay(gameInfo?.app_name, gameInfo.runner)
+    const shouldOpenOverlay =
+      gameInfo &&
+      (gameInfo.runner === 'hyperplay' ||
+        (gameInfo.runner === 'sideload' && gameInfo.web3?.supported))
+
+    if (shouldOpenOverlay) openOverlay(gameInfo?.app_name, gameInfo.runner)
```
Comment (BrettCleary) [inline src/backend/launcher.ts:797]: this solves https://github.com/HyperPlay-Gaming/hyperplay-desktop-client/issues/598

File: src/backend/utils/ipc_renderer_helper.ts
Comment (BrettCleary) [inline src/backend/utils/ipc_renderer_helper.ts:1]: This could replace all of our window.api preload functions that listen for backend changes with `ipcRenderer.on`. It accomplishes the same tasks with less boilerplate

File: src/common/state/interfaces.ts
Comment (BrettCleary) [inline src/common/state/interfaces.ts:1]: inherited by mobx classes

File: src/frontend/ExtensionHandler/index.tsx [hunk 2]
```diff
@@ -82,10 +82,6 @@ const ExtensionHandler = function () {
   }
 
   const bindEthereumListeners = function () {
-    window.addEventListener('message', (event: MessageEvent) => {
```
Comment (BrettCleary) [inline src/frontend/ExtensionHandler/index.tsx:85]: this was spamming the dev tools console and wasn't useful

File: src/frontend/ExtensionManager/index.tsx [hunk 3]
```diff
@@ -26,27 +27,48 @@ const ExtensionManager = function () {
...
+
+  if (OverlayState.title === 'HyperPlay Extension') {
+    mmContainerStyle.left = 0
+    mmContainerStyle.right = 'unset'
+    mmContainerStyle.top = 0
+    // document.documentElement.style.setProperty('--top-navbar-height', `0px`)
+  } else if (
+    OverlayState.title === 'HyperPlay Extension Overlay' ||
+    OverlayState.title === 'HyperPlay Browser Game' ||
+    OverlayState.title === 'HyperPlay Web Game'
```
Comment (BrettCleary) [inline src/frontend/ExtensionManager/index.tsx:40]: setting the window title with electron is flaky/buggy, so we can pass a title string to the frontend mobx class to indicate what type of overlay this is

File: src/frontend/OverlayManager/ExtensionManager/index.tsx
Comment (BrettCleary) [inline src/frontend/OverlayManager/ExtensionManager/index.tsx:1]: This provided the same functionality of `src/frontend/ExtensionManager` for the overlay. Now we can use `src/frontend/ExtensionManager` for both the overlay and the main window

File: src/frontend/OverlayManager/ExtensionToastManager/index.tsx
Comment (BrettCleary) [inline src/frontend/OverlayManager/ExtensionToastManager/index.tsx:1]: This functionality is now covered by `src\frontend\OverlayManager\ToastManager\index.tsx`. Previously we had separate toast managers for extension and external wallets. This is unnecessary and can result in bugs like one toast covering another.

File: src/frontend/OverlayManager/ToastManager/index.tsx
Comment (BrettCleary) [inline src/frontend/OverlayManager/ToastManager/index.tsx:1]: much simpler now :)

File: src/frontend/OverlayManager/index.tsx
Comment (BrettCleary) [inline src/frontend/OverlayManager/index.tsx:1]: this is the overlay. It basically just has flags by each component and lets the mobx class tell it what to show

File: src/frontend/ViewManager.tsx
Comment (BrettCleary) [inline src/frontend/ViewManager.tsx:1]: no more passing the render state in as query params. Now we can change what is shown after the window is created

File: src/frontend/state/DeviceState.tsx
Comment (BrettCleary) [inline src/frontend/state/DeviceState.tsx:1]: this is provided as an option so that other mobx classes can easily consume this. I thought about refactoring the context provider `platform` to use this one, but it touches too many components. It also does not give us any performance benefits as this is a read only value used throughout the whole tree.

File: src/frontend/store/ExtensionStore.ts
Comment (BrettCleary) [inline src/frontend/store/ExtensionStore.ts:1]: we should call our mobx classes `*State.ts` instead of `*Store.ts` because store can also mean:
- hp/gog/epic stores
- electron store
- frontend typed store
- cached backend store

I refactored the mobx classes that I touched with this PR and left the others to be renamed on another PR

File: src/frontend/store/index.tsx [hunk 4]
```diff
@@ -5,7 +5,7 @@ const StoreController = () => {
   useEffect(() => {
     const stores = import.meta.glob<{ default: GenericStore }>([
       './*Store.ts*',
-      '../state/*State.ts'
+      '../state/*State.ts*'
```
Comment (BrettCleary) [inline src/frontend/store/index.tsx:8]: calls the `init()` function for `*State.ts` too

**Comment by nyghtstalker** [comment]
**-- Windows PC - Wallet Locked --**

**General:**

- Scroll bar appears after incorrect PW
image



**Windows game**
- Launch a Windows game (Last Resort) -> trigger overlay right away (before toast) -> toast trigger will close overlay -> 
  open overlay -> overlay is now in a broken state and the mouse will be "under" it
- Trigger overlay -> click forgot PWD link -> blank box opens -> must close overlay and click to show content
image

- Trigger overly -> click MM support link -> blank box -> close overlay -> closes box



**Browser game (Defi Kingdoms)**
- Toast appears **_immediately_**
- Clicking "X" on toast OPENS the overlay
- Closing game via overlay Exit button -> MM login panel open in client
- Screen "blinks" immediately after launching before toast appears
- Full overlay seems to "blink" after the toast fades

**Comment by nyghtstalker** [comment]
**-- MAC - Wallet Locked --**

**General:**

- Scroll bar appears after incorrect PW in overlay

**MAC game (Dark Throne)**
- Launch a MAC game -> trigger overlay right away (before toast) -> toast trigger will close 
   overlay -> open overlay -> overlay is now in a broken state and the mouse will be "under" it
- Trigger overlay -> click forgot PWD link -> Changes screen to client -> box opens under the 
   overlay -> must close overlay to access
- Trigger overly -> click MM support link -> box opens under the overlay -> closing overlay 
   closing the box 
- Toast appears after 10-11s (different than native Windows)

**MAC game (Zoo Racers)** 
- Closing overlay brings the client to the front with the game behind
- Launching overlay during Unity screen will break the overlay -> overlay has a black screen
image


**Browser game (Defi Kingdoms)**
- Toast displays a different font before switching to the correct font
image
image
- Toast appears after 6-7s (different than browser Windows)
- Closing game via overlay Exit button -> MM login panel open in client
Screen "blinks" immediately after launching before toast appears
Full overlay seems to "blink" after the toast fades

File: src/frontend/ExtensionManager/index.tsx [hunk 5]
```diff
@@ -26,27 +27,48 @@ const ExtensionManager = function () {
   const rootRef = useRef<HTMLDivElement>(null)
   const trueAsStr = 'false' as unknown as boolean | undefined
 
+  const mmContainerStyle = {} as React.CSSProperties
+
+  if (OverlayState.title === 'HyperPlay Extension') {
```
Comment (eliobricenov) [inline src/frontend/ExtensionManager/index.tsx:32]: Maybe we can have a `OverlayState.mode`? We could also have the title derived from the mode. We're doing a string comparison here that looks a bit weird

```ts
type OverlayMode = 'extension' | 'browser-game' | 'web-game';

class OverlayState {
    mode: OverlayMode;

    get title(): {
        const titles: Record<OverlayMode, string> = {
            extension: 'HyperPlay Extension Overlay',
            'browser-game': 'HyperPlay Browser Game',
            'web-game': 'HyperPlay Web Game'
        }
        return titles[this.mode];
    }
}
```

File: src/frontend/ExtensionManager/index.tsx [hunk 6]
```diff
@@ -26,27 +27,48 @@ const ExtensionManager = function () {
...
+    mmContainerStyle.left = 0
+    mmContainerStyle.right = 'unset'
+    mmContainerStyle.top = 0
+    // document.documentElement.style.setProperty('--top-navbar-height', `0px`)
```
Comment (red-game-dev) [inline src/frontend/ExtensionManager/index.tsx:36]: We should remove commented code, we can always access that through git version control if we need later 🙌

**Comment by red-game-dev** [comment]
The code looks good and a lot of code was cleaned/refactored, it is more readable and follows the principles. Looks good to me, will aim next to test it 🙌

File: src/frontend/OverlayManager/index.tsx [hunk 7]
```diff
@@ -1,61 +1,120 @@
...
+          {OverlayState.renderState.showHintText ? (
+            <div className={`${BrowserGameStyles.closeOverlayText} title`}>
+              {t('hyperplayOverlay.closeOverlay', {
+                defaultValue:
+                  'Press {{overlayKeyMod}} + X to close the overlay',
+                overlayKeyMod: DeviceState.isMac ? 'Option' : 'Alt'
+              })}
+            </div>
+          ) : null}
+          {OverlayState.renderState.showExitGameButton ? (
```
[the remark this answers is not in the document]
Comment (flavioislima) [reply src/frontend/OverlayManager/index.tsx:74]: I agree, this is confusing. Perhaps split this into smaller components.

**Review by flavioislima — APPROVED** [review]
Left a few comments but looks pretty good.
Did some basic testing and works well.

File: src/frontend/ExtensionManager/index.tsx (hunk 6 above)
Comment (BrettCleary) [reply src/frontend/ExtensionManager/index.tsx:36]: sure. accidentally left it in from debugging

File: src/backend/storeManagers/storeManagerCommon/games.ts [hunk 8]
```diff
@@ -161,6 +171,34 @@ const openNewBrowserGameWindow = async (
     })
 
     browserGame.loadURL(url)
+    // this is electron's suggested way to prevent visual flash
+    // https://github.com/electron/electron/blob/main/docs/api/browser-window.md#using-the-ready-to-show-event
+    browserGame.on('ready-to-show', () => browserGame.show())
```
[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/backend/storeManagers/storeManagerCommon/games.ts:176]: it doesn't completely fix it, but it makes it a little bit better. I also tried https://www.electronjs.org/docs/latest/api/web-contents#event-dom-ready and https://www.electronjs.org/docs/latest/api/web-contents#event-did-finish-load and 'ready-to-show' gave the best behavior

[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/backend/storeManagers/storeManagerCommon/games.ts:176]: iirc 'dom-ready' was the same as before. 'did-finish-load' took 10+ seconds to load and still had white screen for ~1 second

File: src/frontend/state/OverlayState.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+  title = ''
+
+  constructor() {
+    makeAutoObservable(
+      this,
+      {},
+      {
+        deep: true,
+        proxy: false,
```
[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/frontend/state/OverlayState.ts:28]: it makes it easier to debug, a little bit faster and the shape of the OverlayState object doesn't change over time

[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/frontend/state/OverlayState.ts:28]: https://mobx.js.org/observable-state.html#avoid-proxies

```
The second difference is that observable creates a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) object, to be able to trap future property additions in case you use the object as a dynamic lookup map. If the object you want to make observable has a regular structure where all members are known up-front, we recommend to use makeObservable as non proxied objects are a little faster, and they are easier to inspect in the debugger and console.log.
```

File: src/common/types/proxy-types.ts [hunk 10]
```diff
@@ -26,3 +26,8 @@ export async function wait(ms: number) {
     }, ms)
   })
 }
+
+export interface OverlayWindowState {
+  ignoreInput: boolean
+  state?: 'show' | 'hide' | null
```
[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/common/types/proxy-types.ts:32]: it needs to switch from null to hide to pass the 'hide' value from the backend to the frontend. if it's 'hide' by default and then we tell it to 'hide', it will not send to the frontend because it's the same value

[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/common/types/proxy-types.ts:32]: I don't think we can init to undefined because we have to set up a listener to it. I don't think that the observable will be observed if it's undefined

File: src/common/typedefs/ipcBridge.d.ts [hunk 11]
```diff
@@ -204,6 +210,11 @@ interface HyperPlayAsyncIPCFunctions {
     channelId: number,
     accessCode: string
   ) => Promise<LicenseConfigValidateResult>
+  get_wallet_state_address: () => Promise<string>
```
[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/common/typedefs/ipcBridge.d.ts:213]: personally, I think it is easier to see the distinction between state object, update/action/get initial call, and property name with delimiters. we have to differentiate these all in the same namespace as a string so we need to make sure there are no possible collisions. this is not as immediately clear by visual inspection when using camel case

these are not meant to be called generally from the frontend either. They are meant only to update the frontend mobx state. a different naming convention does clearly distinguish this system from the rest of the app even if it's unconventional

File: src/frontend/screens/TransactionNotification/index.tsx [hunk 12]
```diff
@@ -2,38 +2,32 @@ import React from 'react'
...
-    item.state === TransactionState.CONFIRMED ||
-    item.state === TransactionState.FAILED
-  ) {
-    setTimeout(() => transactionStore.closeTransaction(item.id), 5000)
-  }
+    item === null ||
+    !item.isOpen ||
+    WalletState.provider === PROVIDERS.METAMASK_EXTENSION ||
+    WalletState.provider === PROVIDERS.UNCONNECTED
+  )
```
[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/frontend/screens/TransactionNotification/index.tsx:18]: i will change this but as a sidenote, if this is important to you guys @eliobricenov @flavioislima, you need to enable eslint curly and fix all the errors https://github.com/HyperPlay-Gaming/hyperplay-desktop-client/issues/581

it is a waste of our time to discuss eslint style errors on PR's

File: src/frontend/screens/TransactionNotification/components/PhoneIcon/index.tsx [hunk 13]
```diff
@@ -5,24 +5,25 @@ import {
   PhoneCompletedIcon,
   PhoneWaitingConfimationSpinner
 } from 'frontend/assets/hyperplay'
-import { TransactionState } from 'frontend/store/types'
 import Loading from './Loading'
+import { statusType } from '@hyperplay/ui/dist/components/TransactionToasts'
```
[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/frontend/screens/TransactionNotification/components/PhoneIcon/index.tsx:9]: yeah we can do that. it will need a different name though and we'll need to keep the old export until everyone is on the new ui version

File: src/common/typedefs/ipcBridge.d.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/common/typedefs/ipcBridge.d.ts:213]: also naming conventions are usually used to differentiate between things like classes, objects, files, etc. that you don't necessarily know the difference between by just looking at the line it's used. For window.api calls, they are all prefaced by `window.api.` which makes it pretty clear what's going on. so quite frankly it doesn't matter much for the stylistic benefit

[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/common/typedefs/ipcBridge.d.ts:213]: as a broader point, we use camel case for pretty much everything and our style does not clearly distinguish a lot of these symbols, so we're not getting much benefit out of that

File: src/frontend/OverlayManager/index.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/frontend/OverlayManager/index.tsx:74]: @eliobricenov sure I can make that change. it is ultimately a stylistic preference

@flavioislima with elio's method of assigning it to a variable, the variable name serves much the same purpose of a different component. I'll just go with that since there is no other benefit to breaking out into a separate component

File: src/frontend/ExtensionManager/index.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/frontend/ExtensionManager/index.tsx:32]: i feel like this is a lot of added complexity for what is still basically a case of string matching. it also only needs to match in two places in the code

I like your idea about the OverlayMode type though. I will add that and change the title type to use that

File: src/frontend/screens/TransactionNotification/components/PhoneIcon/index.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (BrettCleary) [reply src/frontend/screens/TransactionNotification/components/PhoneIcon/index.tsx:9]: PR here https://github.com/HyperPlay-Gaming/hyperplay-ui/pull/103

[the remark this answers is not in the document]
Comment (eliobricenov) [reply src/frontend/screens/TransactionNotification/components/PhoneIcon/index.tsx:9]: approved

File: src/frontend/OverlayManager/index.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (eliobricenov) [reply src/frontend/OverlayManager/index.tsx:74]: I think multiple components will make even more noise. This is not a blocker by any chance.

File: src/frontend/ExtensionManager/index.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (eliobricenov) [reply src/frontend/ExtensionManager/index.tsx:32]: The map is only to leverage typescript in case we add a different mode, we'd be forced to add a title to it (it has happened to me, I forge to add the switch case and the result is an empty string).

Not a blocker.

File: src/common/types/proxy-types.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (eliobricenov) [reply src/common/types/proxy-types.ts:32]: Interesting, in that case maybe the state should not be optional but nullable.

File: src/common/typedefs/ipcBridge.d.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (eliobricenov) [reply src/common/typedefs/ipcBridge.d.ts:213]: I don't necessarily love camel case but I think having consistency in the code is good. Even if that means using snake case for some parts of the code.

Either way, this is a nitpick and not a blocker.

**Review by eliobricenov — APPROVED** [review]
lgtm, great work ser. Thank you for answering my questions.

**Review by red-game-dev — APPROVED** [review]
Very good work, it's awesome we got so many new states now that we can utilize 🙌 Cheers 🍻

**Review by nyghtstalker — APPROVED** [review]
One big PR ready to go! Let's do this!!!
